onnxruntime/core/mlas/lib/aarch64/DepthwiseQConvKernelSize9Neon.S

```diff
@@ -0,0 +1,366 @@
+        mov     x14, x24
+        mov     x15, x25
+        mov     x16, x26
+        mov     x17, x27
+        mov     x19, x28
+
+.LMlasConvSymDepthwiseKernelSize9_Loaded_Input:
+
+        eor     x8, x8, x8                  // Processed channels
+        umov    x1, v12.D[0]                // filter
+        umov    x5, v13.D[0]                // bias
+        umov    x7, v13.D[1]                // scale
+    
+        cmp     x8, x2                      // Save one register by not using count down to zero here
+        bhs     .LMlasConvSymDepthwiseKernelSize9_Finish_Channels16_Loop
+
+.LMlasConvSymDepthwiseKernelSize9_Channels16_Loop:
+        ld1     {v10.16b}, [x1], x2         // vk0
+        ldr     q16, [x20, x8]              // out0 vi0
+        ldr     q17, [x10, x8]              // out1 vi0
+        ld1     {v6.4s, v7.4s, v8.4s, v9.4s}, [x5], #64       // bias vacc 0-15 for outs
+        ld1     {v11.16b}, [x1], x2         // vk1
+        ldr     q18, [x21, x8]              // out0 vi1
+        ldr     q19, [x11, x8]              // out1 vi1
+
+        eor     v16.16b, v16.16b, v5.16b    // -128 to signed int8
+        eor     v17.16b, v17.16b, v5.16b
+        eor     v18.16b, v18.16b, v5.16b
+        eor     v19.16b, v19.16b, v5.16b
+
+        smull   v24.8h, v10.8b, v16.8b
+        smull2  v25.8h, v10.16b, v16.16b
+        smull   v26.8h, v10.8b, v17.8b
+        smull2  v27.8h, v10.16b, v17.16b
+        smull   v28.8h, v11.8b, v18.8b
+        smull2  v29.8h, v11.16b, v18.16b
+        smull   v30.8h, v11.8b, v19.8b
+        smull2  v31.8h, v11.16b, v19.16b
+
+
+        ld1     {v14.16b}, [x1], x2         // vk2
+        ldr     q20, [x22, x8]              // out0 vi2
+        ldr     q21, [x12, x8]              // out1 vi2
+        ld1     {v15.16b}, [x1], x2         // vk3
+        ldr     q22, [x23, x8]              // out0 vi3
+        ldr     q23, [x13, x8]              // out1 vi3
+
+        eor     v20.16b, v20.16b, v5.16b
+        eor     v21.16b, v21.16b, v5.16b
+        eor     v22.16b, v22.16b, v5.16b
+        eor     v23.16b, v23.16b, v5.16b
+
+        smlal   v24.8h, v14.8b, v20.8b
+        smlal2  v25.8h, v14.16b, v20.16b
+        smlal   v26.8h, v14.8b, v21.8b
+        smlal2  v27.8h, v14.16b, v21.16b
+        smlal   v28.8h, v15.8b, v22.8b
+        smlal2  v29.8h, v15.16b, v22.16b
+        smlal   v30.8h, v15.8b, v23.8b
+        smlal2  v31.8h, v15.16b, v23.16b
+
+        saddw   v16.4s, v6.4s, v24.4h       // dup acc for out1
+        saddw2  v17.4s, v7.4s, v24.8h
+        saddw   v18.4s, v8.4s, v25.4h
+        saddw2  v19.4s, v9.4s, v25.8h
+
+        saddw   v6.4s, v6.4s, v26.4h
+        saddw2  v7.4s, v7.4s, v26.8h
+        saddw   v8.4s, v8.4s, v27.4h
+        saddw2  v9.4s, v9.4s, v27.8h
+
+        saddw   v16.4s, v16.4s, v28.4h
+        saddw2  v17.4s, v17.4s, v28.8h
+        saddw   v18.4s, v18.4s, v29.4h
+        saddw2  v19.4s, v19.4s, v29.8h
+
+        ld1     {v10.16b}, [x1], x2         // vk4
```

Comment:
since you are already moving these loads before a couple of saddw, why not move them even earlier to give you more safety margin? I think you may safely leave 12 cycles between memory load and its usage, so that loading from l2 cache won't cause any delay.  

More over, you might want to interleave load with saddw. This is because in A75 CPUs, load and saddw uses different pipelines and can be issued concurrently.

Reply:
Nice.